### PR TITLE
FEAT: Make LoRA work with custom models

### LINKED ISSUE #365: GPT2 Models Merge and Unload
Hello again, do you think about merging for gpt2 models? It would be great if you could do it. When we can't test new models (Alpaca etc), we have to use the old ones (GPT-2). Do you have another method ? I wish you a good work

```

from peft import PeftModel, PeftConfig

peft_model_id = "/content/PEFT/V_1.0/"

config = PeftConfig.from_pretrained(peft_model_id)

tokenizer = AutoTokenizer.from_pretrained(config.base_model_name_or_path)

model = AutoModelWithLMHead.from_pretrained(
    config.base_model_name_or_path, 
    #torchscript = True,
    #load_in_8bit=True,
    #device_map = "auto",
    use_cache = False
)
print("Base Model : ", model.device)

device = "cuda:0" if torch.cuda.is_available() else "cpu"
device = "cpu"

model = PeftModel.from_pretrained(model, peft_model_id)
print("Peft Model : " ,model.device)

if torch.__version__.startswith("2.0"): 
    print("Model Compile Using...")   
    model_comp = torch.compile(model)

model = model.merge_and_unload()

│ in <cell line: 1>:1                                                                              │
│                                                                                                  │
│ /usr/local/lib/python3.9/dist-packages/peft/tuners/lora.py:305 in merge_and_unload               │
│                                                                                                  │
│   302 │   │   as a standalone model.                                                             │
│   303 │   │   """                                                                                │
│   304 │   │   if getattr(self.config, "model_type", None) == "gpt2":                             │
│ ❱ 305 │   │   │   raise ValueError("GPT2 models are not supported for merging LORA layers")      │
│   306 │   │                                                                                      │
│   307 │   │   if getattr(self.model, "is_loaded_in_8bit", False):                                │
│   308 │   │   │   raise ValueError("Cannot merge LORA layers when the model is loaded in 8-bit   │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ValueError: GPT2 models are not supported for merging LORA layers

### PR DESCRIPTION
**Pull request by BenjaminBossan**
This is a basis for discussion and not yet ready to be merged.

There are surprisingly few changes necessary to make LoRA work with custom models. Those changes mostly revolve around attributes or methods that are present by default on transformers models but not necessarily on custom models, and many of those changes are only required for tests. This PR identifies those lines of code and fixes them so that they work with custom models _without_ creating any backwards incompatibility for models that are already supported.

A new suite of tests is introduced that checks that LoRA works for various kinds of custom models.

In `_test_inference_safetensors`, I removed the `layers_to_transform` config setting, as layer indexing does not work with many custom models.

There are still some TODOs for this PR, or possible future PRs:

- Some more tests need to be added
- Some tests (those using conv layers) are failing and I'm not sure why, take a look at `tests_failing.py`
- It could be useful to have an example and/or update the docs.
- Layer indexing doesn't really work with custom models, except if they have a very specific naming pattern. I'll do a small refactor of layer indexing in a separate PR, perhaps this also enable it to work better with custom models.
- Test if custom models also work with AdaLora (and IA³).

Also fixes #365

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
This overall looks great to me and very clean, I really like the new approach of creating a default dummy model config
Let's maybe add few lines on the documentation to explain what it means that LoRA is supported for any model with a handy example snippet 
Made also: https://github.com/BenjaminBossan/peft/pull/1 that should unlock the issues with Conv1D layers

**Comment by BenjaminBossan** [comment]
> Let's maybe add few lines on the documentation to explain what it means that LoRA is supported for any model with a handy example snippet

Yes, I'll work on that.

> Made also: [BenjaminBossan#1](https://github.com/BenjaminBossan/peft/pull/1) that should unlock the issues with Conv1D layers

Great, this fixed the issues in `test_failing.py`. There are some more failing tests when conv layers are used, but they're probably unrelated, I'll investigate and get back once I know what's going on.

File: tests/testing_common.py [hunk 1]
```diff
@@ -284,7 +285,7 @@ def _test_merge_layers(self, model_id, config_cls, config_kwargs):
         if config.peft_type != "LORA":
             with self.assertRaises(AttributeError):
                 model = model.merge_and_unload()
-        elif model.config.model_type == "gpt2":
+        elif model.config["model_type"] == "gpt2":
```
Comment (pacman100) [inline tests/testing_common.py:288]: should be resolved by the Conv1D support you have added, so no need for this.

**Review by pacman100 — COMMENTED** [review]
This is awesome work @BenjaminBossan to enable LoRA with custom models 🚀! Let's have it as part of the release as it also solves a long standing issue with GPT-2 not working with `merge_and_unload()` as it was using Conv1D. 

Left a couple comments

File: src/peft/tuners/lora.py [hunk 2]
```diff
@@ -449,7 +458,10 @@ def merge_and_unload(self):
                     )
                 else:
                     bias = target.bias is not None
-                    new_module = torch.nn.Linear(target.in_features, target.out_features, bias=bias)
+                    if getattr(target, "is_target_conv_1d_layer", False):
+                        new_module = Conv1D(target.out_features, target.in_features)
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/tuners/lora.py:462]: Indeed, I'll remove it.

File: tests/testing_common.py (hunk 1 above)
Comment (BenjaminBossan) [reply tests/testing_common.py:288]: Yes, this works now for LoRA, unfortunately not for IA³. Not sure why, but it didn't look like a quick fix to me.

**Review by younesbelkada — APPROVED** [review]
Looking great, thank you !

**Review by younesbelkada — COMMENTED** [review]
Can you add few lines in the documentation about this feature? It could live in the lora section. It can be also done in a follow up PR

**Comment by BenjaminBossan** [comment]
> Can you add few lines in the documentation about this feature? It could live in the lora section. It can be also done in a follow up PR

Yes, I'd like to add some docs and examples. Still, I would like to ensure that everything works well before merging, so my preference is to deliver the docs and examples first, as new issues could surface. That means it's not going to be ready if we want to release today, but I think that's fine.

**Review by pacman100 — APPROVED** [review]

**Review by pacman100 — APPROVED** [review]
Hello, Thank you @BenjaminBossan for all the awesome work in making LoRA work with custom models, super cool! 🤗

**Comment by tuntun990606** [comment]
> > Can you add few lines in the documentation about this feature? It could live in the lora section. It can be also done in a follow up PR
> 
> Yes, I'd like to add some docs and examples. Still, I would like to ensure that everything works well before merging, so my preference is to deliver the docs and examples first, as new issues could surface. That means it's not going to be ready if we want to release today, but I think that's fine.

May I ask if the current model can be used for custom models? After fine-tuning using the custom model with the addition of Lora module, the model still reports an error in dropout

**Comment by BenjaminBossan** [comment]
> May I ask if the current model can be used for custom models?

It should work now, here is an example notebook: https://github.com/BenjaminBossan/peft/blob/DOC-custom-models/examples/multilayer_perceptron/multilayer_perceptron_lora.ipynb

> After fine-tuning using the custom model with the addition of Lora module, the model still reports an error in dropout

Could you provide some code to produce that error?

**Comment by tuntun990606** [comment]
> This is a basis for discussion and not yet ready to be merged.
> 
> There are surprisingly few changes necessary to make LoRA work with custom models. Those changes mostly revolve around attributes or methods that are present by default on transformers models but not necessarily on custom models, and many of those changes are only required for tests. This PR identifies those lines of code and fixes them so that they work with custom models _without_ creating any backwards incompatibility for models that are already supported.
> 
> A new suite of tests is introduced that checks that LoRA works for various kinds of custom models.
> 
> In `_test_inference_safetensors`, I removed the `layers_to_transform` config setting, as layer indexing does not work with many custom models.
> 
> There are still some TODOs for this PR, or possible future PRs:
> 
> * Some more tests need to be added
> * Some tests (those using conv layers) are failing and I'm not sure why, take a look at `tests_failing.py`
> * It could be useful to have an example and/or update the docs.
> * Layer indexing doesn't really work with custom models, except if they have a very specific naming pattern. I'll do a small refactor of layer indexing in a separate PR, perhaps this also enable it to work better with custom models.
> * Test if custom models also work with AdaLora (and IA³).
> 
> Also fixes #365
 

> 

 I would like to use PEFT to fine tune the Adaspech model, but I noticed that after adding PEFT code, print (model. base_model. model. module. speaker_emb. ora_embedding. A) outputs all 0 matrices, which means that this part did not participate in training. This is part of my code. May I ask where the problem is with me?

`
# Prepare model
    model, optimizer = load_pretrain(args, configs, device, train=True)
    model = nn.DataParallel(model)
    model = model.to(device)
    peft_config = peft.LoraConfig(
        r=8,
        target_modules=["module.speaker_emb"])
    model = peft.get_peft_model(model, peft_config)
    model.print_trainable_parameters()
    num_param = get_param_num(model)
    Loss = AdaSpeechLoss(preprocess_config, model_config).to(device)
`

when i input this code：
‘print(model.base_model.model.module.speaker_emb.lora_embedding_A) ’
it always output that：
parameterDict(  (default): Parameter containing: [torch.cuda.FloatTensor of size 8x174 (GPU 0)])   
Its value will not be updated

**Pull request merged**